designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,808 @@
+    * `hasRule(ruleId)` - determine if the given rule is in the core
+* `cwd` - the current working directory for ESLint (might be different than `process.cwd()` but always matches `CLIEngine.options.cwd`, see https://github.com/eslint/eslint/issues/11218)
+
+This information allows users to make logical decisions about how the config should be constructed.
+
+A configuration function may return an object or an array of objects. An error is thrown if any other type of value is returned.
+
+#### Checking for Rule Existence
+
+One of the problems with shareable configs today is when a new rule is added to the ESLint core, shareable configs using that rule are not valid for older versions of ESLint (because ESLint validates that configured rules are present). With advanced configs, a shareable config could detect if a new rule is present before deciding to include it, for example:
+
+```js
+const requireIndex = require("requireindex");
+
+exports.config = (context) => {
+    const myConfig = {
+        files: ["*.js"],
+        defs: {
+            ruleNamespaces: {
+                custom: requireIndex("./custom-rules")
+            },
+        }
+        rules: {
+            "custom/my-rule": "error"
+        }
+    };
+
+    if (context.hasRule("some-new-rule")) {
+        myConfig.rules["some-new-rule"] = ["error"];
+    }
+
+    return myConfig;
+};
+```
+
+#### Including Function Configs in an Array
+
+A function config can be used anywhere a config object or a config array is valid. That means you can insert a function config as a config array member:
+
+```js
+exports.config = [
+    (context) => someObject,
+    require("eslint-config-myconfig")
+];
+```
+
+Each function config in an array will be executed with a `context` object when ESLint evaluates the configuration file. This also means that shareable configs can export a function instead of an object or array.
+
+**Note:** If a function config inside of a config array happens to return an array, then those config array items are flattened as with any array-in-array situation.
+
+### Configuration Location Resolution
+
+When ESLint is executed, the following steps are taken to find the `eslint.config.js` file to use:
+
+1. If the `-c` flag is used then the specified configuration file is used. There is no further search performed.
+1. Otherwise:
+    a. Look for `eslint.config.js` in the current working directory. If found, stop searching and use that file.
+    b. If not found, search up the directory hierarchy looking for `eslint.config.js`.
+    c. If a `eslint.config.js` file is found at any point, stop searching and use that file.
+    d. If `/` is reached without finding `eslint.config.js`, then stop searching and output a "no configuration found" error.
+
+This approach will allow running ESLint from within a subdirectory of a project and get the same result as when ESLint is run from the project's root directory (the one where `eslint.config.js` is found).
```

Comment:
This seems like a nice solution to the issue of always having to have the root directory of a repository open as the root in the file tree for editor integrations to work.